### ISSUE TITLE
Allow to set --node-ip in kubelet.service

### ISSUE THREAD

**Issue by FlorinPeter** [openshift/machine-config-operator#944]
**Description**

We are facing currently some issues on AWS when adding additional network interfaces to a master or worker node. Some time it happens that the first InternalIP is not the IP from the first network interface.
To avoid this issue we tried to add the --node-ip parameter by overwriting the kubelet.service as described here https://github.com/openshift/installer/blob/master/docs/user/customization.md but it it turns out that this customization is not longer possible.

**Steps to reproduce the issue:**
1. Install cluster on AWS
2. add additional network interfaces

**Describe the results you received:**
Node ip-192-168-43-78.eu-central-1.compute.internal is reporting: "during
      bootstrap: unexpected on-disk state validating against rendered-worker-9480d1bae1f38b6a6efcbd8152249e7e"

**Describe the results you expected:**
A possibility to overwrite the kubelet.service or a possibility set the --node-ip parameter via environments variables 

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `oc adm release info --commits | grep machine-config-operator`:**

```
 machine-config-controller                     https://github.com/openshift/machine-config-operator                       02c07496ba0417b3e12b78fb32baf6293d314f79
  machine-config-daemon                         https://github.com/openshift/machine-config-operator                       02c07496ba0417b3e12b78fb32baf6293d314f79
  machine-config-operator                       https://github.com/openshift/machine-config-operator                       02c07496ba0417b3e12b78fb32baf6293d314f79
  machine-config-server                         https://github.com/openshift/machine-config-operator                       02c07496ba0417b3e12b78fb32baf6293d314f79
  setup-etcd-environment                        https://github.com/openshift/machine-config-operator                       02c07496ba0417b3e12b78fb32baf6293d314f79
```

**Additional environment details (platform, options, etc.):**

**Comment by runcom** [comment]
This probably needs to live in the kubelet Config? @rphillips

**Comment by FlorinPeter** [comment]
The current workaround we found is to overwrite the https://github.com/openshift/machine-config-operator/blob/master/templates/master/01-master-kubelet/_base/units/kubelet.yaml but this only for testing purpose. 
```yaml
name: "kubelet.service"
enabled: true
contents: |
  [Unit]
  Description=Kubernetes Kubelet
  Wants=rpc-statd.service crio.service
  After=crio.service
  [Service]
  Type=notify
  ExecStartPre=/bin/mkdir --parents /etc/kubernetes/manifests
  ExecStartPre=/bin/rm -f /var/lib/kubelet/cpu_manager_state
  EnvironmentFile=/etc/os-release
  EnvironmentFile=-/etc/kubernetes/kubelet-workaround
  EnvironmentFile=-/etc/kubernetes/kubelet-env
  ExecStart=/usr/bin/hyperkube \
      kubelet \
        --config=/etc/kubernetes/kubelet.conf \
        --bootstrap-kubeconfig=/etc/kubernetes/kubeconfig \
        --rotate-certificates \
        --kubeconfig=/var/lib/kubelet/kubeconfig \
        --container-runtime=remote \
        --container-runtime-endpoint=/var/run/crio/crio.sock \
        --node-labels=node-role.kubernetes.io/master,node.openshift.io/os_id=${ID} \
        --minimum-container-ttl-duration=6m0s \
        --cloud-provider={{cloudProvider .}} \
        --volume-plugin-dir=/etc/kubernetes/kubelet-plugins/volume/exec \
        {{cloudConfigFlag . }} \
        --register-with-taints=node-role.kubernetes.io/master=:NoSchedule \
        --v=3 \
        --node-ip=${KUBELET_NODE_IP} \
        
  Restart=always
  RestartSec=10
  [Install]
  WantedBy=multi-user.target
```

We can populate the $KUBELET_NODE_IP via drop-in /etc/systemd/system/kubelet.service.d/node-ip.conf by creating a MachineConfig
/etc/systemd/system/kubelet.service.d/node-ip.conf
```
ExecStartPre=/bin/bash -c "/bin/systemctl set-environment KUBELET_NODE_IP=$(hostname --ip-address)" 
```

Another option would be to have at the end of the kubelet command a var like $KUBELET_ARGS that can be populated via /etc/kubernetes/kubelet-workaround or /etc/kubernetes/kubelet-env

**Comment by cgwalters** [comment]
See also https://github.com/openshift/machine-config-operator/issues/763 (which morphed into a more specific solution) but brought up the general kubelet argument problem.

**Comment by cgwalters** [comment]
Let's go to the toplevel problem though:

> Some time it happens that the first InternalIP is not the IP from the first network interface.

By "first" here you mean...the interface for the installer-created network?  (Is this an IPI or UPI install?)  What are you doing with the other networks you've created?

This seems like a problem better solved automatically rather than having you work around it.

**Comment by dcbw** [comment]
@cgwalters @FlorinPeter you want this backported to OpenShift to magically fix all your InternalIP problems: https://github.com/kubernetes/kubernetes/pull/79391

**Comment by dcbw** [comment]
I don't think we really want to allow --node-ip to be set. Let's keep that dynamic and fix the real bug, which was non-deterministic ordering of InternalIPs when posted to the Kube apiserver.

**Comment by dcbw** [comment]
The InternalIP ordering issue has been fixed in origin master in https://github.com/openshift/origin/pull/23345

**Comment by FlorinPeter** [comment]
@dcbw nice 👍 
Thx for the information. I was not aware that there is an upstream fix.
I guess it will also backported to 4.1?

**Comment by dcbw** [comment]
@FlorinPeter that was not currently the plan as the only user that ran into it worked around the issue and nobody else has ever reported it. However, that could perhaps be done, cc-ing @danwinship for more info.

**Comment by FlorinPeter** [comment]
@dcbw we have this issues since 3.x as we use a second network interface to send logs to a central service. In 3.x we solved this behavior with the --node-ip flag, so for us it is crucial to get this fixed in 4.1.

**Comment by danwinship** [comment]
Yeah, we had been thinking that this didn't need a backport since it was an old bug that hadn't been reported much, but now it seems like that's just because people used to be able to work around it. So we may need to think about backporting it. (But the fix is kind of ugly, and has a lot of potential to break things if it's buggy, so I know the release team is going to be hesitant.)

**Comment by sjenning** [comment]
fix for 4.1
https://github.com/openshift/origin/pull/23367

**Comment by FlorinPeter** [comment]
@danwinship can you please take a look at https://bugzilla.redhat.com/show_bug.cgi?id=1734319
